metrics/resource_view.go

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"go.opencensus.io/metric/metricdata"
+	"go.opencensus.io/resource"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+type storedViews struct {
+	views []*view.View
+	lock  sync.Mutex
+}
+
+type meterExporter struct {
+	m view.Meter    // NOTE: DO NOT RETURN THIS DIRECTLY; the view.Meter will not work for the empty Resource
+	o stats.Options // Cache the option to reduce allocations
+	e view.Exporter
+}
+
+// ResourceExporterFactory provides a hook for producing separate view.Exporters
+// for each observed Resource. This is needed because OpenCensus support for
+// Resources is a bit tacked-on rather than being a first-class component like
+// Tags are.
+type ResourceExporterFactory func(*resource.Resource) (view.Exporter, error)
+type meters struct {
+	meters  map[string]*meterExporter
+	factory ResourceExporterFactory
+	// Cache of Resource pointers from metricskey to Meters, to avoid
+	// unnecessary stringify operations
+	resourceToKey map[*resource.Resource]string
+	lock          sync.Mutex
+}
+
+// Lock regime: lock allMeters before resourceViews. The critical path is in
+// optionForResource, which must lock allMeters, but only needs to lock
+// resourceViews if a new meter needs to be created.
+var resourceViews = storedViews{}
+var allMeters = meters{
+	meters:        map[string]*meterExporter{"": &defaultMeter},
+	resourceToKey: map[*resource.Resource]string{nil: ""},
+}
+
+// RegisterResourceView is similar to view.Register(), except that it will
+// register the view across all Resources tracked by the system, rather than
+// simply the default view.
+func RegisterResourceView(views ...*view.View) error {
+	allMeters.lock.Lock()
```

Comment:
Not sure how much this matters, but from performance point of view you can
Move error list creation outside of locks, or given that you care about _an error_, just ignore the list and have `var rErr error` and on each failure assign `rErr = err` and return `rErr`.

Do we actually need to clone the `view` objects?

Reply:
I think the race detector was unhappy with multiple meters (each with their own background thread) calling `view.canonicalize()` in `go.opencensus.io/stats/view/worker_commands.go`), and nothing in that API made a defensive copy.